### PR TITLE
Support for adjacency matrices in autoregressive models

### PR DESCRIPTION
**Pull request by adrianjav**
This is an implementation of #53, in which I add support to provide an adjacency matrix to the autoregressive models, instead of using the ordering between variables.

I have added some asserts to make sure I cover all corner cases (e.g., providing both an ordering and an adjacency matrix), and added tests to make sure that the zero-entries of the adjacency matrix are indeed zero when computing the Jacobian matrix.

I might have missed something, and I am happy to fix it if that's the case.

PS: I read the policy about commit messages a bit too late, we can merge all commits into a single one and write an appropiate commit message.

### CODE REVIEW

**Comment by francois-rozet** [comment]
Hello @adrianjav, thank you for this PR. I quickly went through the modifications, I'll do a proper review this week. I have a question.

Enabling to construct a `MaskedAutoregressiveTransform` directly from an adjacency matrix makes perfect sense sense. However, I am not sure if a series of masked autoregressive transformations with the same adjacency matrix makes sense. So I am not sure if `MAF` (and `NAF`) should be modified.

 about modifying `MAF` (

**Comment by adrianjav** [comment]
Hi @francois-rozet, thanks for taking a look already to the PR!

> Could you explain in what case having several transformations with the same adjacency would be used?

While we chose to have a single flow layer, we consider these types of architectures in our [article](https://arxiv.org/abs/2306.05415) (Table 1), and the idea is that you have a finer access to the way the information flows through the network.

I implemented this way so that it was direct to use the adjacency matrix on all the derivatives flows (i.e., MAF, NSF, ...) without having to add any more code, but I agree that it can be seen as rather arbitrary. I would propose two alternatives:
1. Add a check so that if the adjacency matrix is used, then the number of layers needs to be 1.
2. Allow to specify either an adjacency matrix or a list of them (one per layer), so that the user has even finer control on the way the flows are structured.

What do you think?

**Comment by francois-rozet** [comment]
Hello @adrianjav,

The current philosophy of Zuko regarding pre-built flows (MAF, NSF, NAF, ...) is to make them very simple and avoid overloading them with many features. We don't want to confuse beginners. If the user needs more control, they are expected to [build their own custom architecture](https://zuko.readthedocs.io/stable/tutorials/basics.html#custom-architecture) using lower-level components such as transformations and distributions.

In this case, we would expect the user to instantiate the list of `MaskedAutoregressiveTransform` themselves given the adjacency matrices and then combine the transformations as a `Flow`. Note that this would be necessary anyway to build a generative (instead of abductive) flow.

```python
flow = zuko.flows.Flow(
    transform=MaskedAutoregressiveTransform(
        features=5,
        context=8,
        adjacency=adjacency,  # 5x5 matrix
    ),
    base=base,
)
```

So I think that this PR should only modify the `MaskedAutoregressiveTransform`. In addition,

- There are constraints on the adjacency $A$ for the transformation to be invertible. I think checking that the adjacency satisfies these constraints would be great.

- The number of "passes" for the inverse transformation can be inferred from the adjacency $A$. If there is a computationally efficient way to do it, it would also be a nice addition.

**Comment by adrianjav** [comment]
Hi, 

Yeah, I fully understand and respect that philosophy. I will revert the changes I made to MAF and keep them local to `MaskedAutoregressiveTransform`. As for the other two points you suggested:

- The matrix $A + I$ should be invertible, not $A$, right? (if we assume a zeroed diagonal).
- If I understand correctly, the number of passess is just the diameter of the graph defined by $A$, right? I can add code to compute that, but it only saves on unecessary computations, isn't it?

I will add the necessary checks for the adjacency matrix.

**Comment by adrianjav** [comment]
I have added the necessary changes and reverted those involving MAF.

As a summary, now it checks whether the matrix has cycles, and infers the order (for the string representation) as well as the number of passes (diameter of the graph). There is no need to check invertibility, since we assume ones in the diagonal and it has no cycles.

For the algorithm, I have simplified the code from the networkx package to deal with our specific case. 

Let me know if I missed something.

File: zuko/flows/autoregressive.py [hunk 1]
```diff
@@ -102,33 +107,87 @@ def __init__(
...
         if passes is None:
             passes = features
+        self.passes = min(max(passes, 1), features)
+
+        assert (order is None) or (
+            adjacency is None
+        ), "Parameters `order` and `adjacency_matrix` are mutually exclusive."
+        assert (passes == features) or (
+            adjacency is None
+        ), "When using `adjacency_matrix`, `passes` has to be the number of features."
```
[the remark this answers is not in the document]
Comment (adrianjav) [reply zuko/flows/autoregressive.py:117]: Could you expand on this? I am not sure I understand what you mean

File: zuko/flows/autoregressive.py [hunk 2]
```diff
@@ -102,33 +107,87 @@ def __init__(
...
+                    indegree_map[child] -= 1
+                    if indegree_map[child] == 0:
+                        zero_indegree.append(child)
+                        del indegree_map[child]
+            all_generations.append(this_generation)
+
+        assert not indegree_map, "The adjacency matrix contains cycles"
+
+        self.passes = len(all_generations)  # Graph diameter
+        self.order = torch.tensor(reduce(list.__add__, all_generations))
```
[the remark this answers is not in the document]
Comment (adrianjav) [reply zuko/flows/autoregressive.py:172]: The adjacency matrix is describing the DAG that the data-generating process follows, so any of its topological orders are indeed the generation order (i.e., if you group those variables that are generated simultaneously), isn't it? In the case of providing the ordering to the constructor, there is no grouping to be made (the adj. matrix is a full lower triangular matrix), so there is only one topological order. Or did I get something wrong?

File: zuko/flows/autoregressive.py [hunk 3]
```diff
@@ -102,33 +107,87 @@ def __init__(
...
+            this_generation = zero_indegree
+            zero_indegree = []
+            for node in this_generation:
+                for child in adjacency[:, node].nonzero():
+                    child = child.item()
+                    indegree_map[child] -= 1
+                    if indegree_map[child] == 0:
+                        zero_indegree.append(child)
+                        del indegree_map[child]
+            all_generations.append(this_generation)
```
[the remark this answers is not in the document]
Comment (adrianjav) [reply zuko/flows/autoregressive.py:167]: Sure, but I think then the assert right below should change, right? To assert that the graph is acyclic.

File: zuko/flows/autoregressive.py [hunk 4]
```diff
@@ -102,33 +107,87 @@ def __init__(
...
+            self.order = torch.div(order, ceil(features / self.passes), rounding_mode='floor')
 
-        if order is None:
-            order = torch.arange(features)
+            in_order = torch.cat((self.order, torch.full((context,), -1)))
+            out_order = torch.repeat_interleave(self.order, self.total)
+            adjacency = out_order[:, None] > in_order
         else:
-            order = torch.as_tensor(order)
+            adjacency.mul_(  # Remove the diagonal (if it is non-zero)
```
[the remark this answers is not in the document]
Comment (adrianjav) [reply zuko/flows/autoregressive.py:131]: I opted for doing it silently since "in reality" the diagonal is non-zero (because of the transformer), and it needs to be made zero for the conditioner. I figured it out that fixing it would be less confusing, but I can also just raise an error.

File: zuko/flows/autoregressive.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (francois-rozet) [reply zuko/flows/autoregressive.py:172]: I think you misunderstood my comment. Currently, the attribute `self.order` is not a topological order so you should not assign a topological order to it.

For example, when `passes=2` and `features=5`, `self.order` is `[0,0,0,1,1]`, meaning that the 3 first features can be computed together (they are in the same generation).

File: zuko/flows/autoregressive.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (francois-rozet) [reply zuko/flows/autoregressive.py:131]: I see what you mean, but then we should impose that the matrix has ones on the diagonal, and raise a error if it has not.

The reason why I prefer raising an error is that if a user makes a mistake (e.g. they permute the rows but they forget to permute the columns), they might not see it if we "fix" it silently.

File: zuko/flows/autoregressive.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (francois-rozet) [reply zuko/flows/autoregressive.py:117]: In the docstring, it is said that `passes` and `order` are **ignored** when `adjacency` is provided. This is not the case here.

[the remark this answers is not in the document]
Comment (adrianjav) [reply zuko/flows/autoregressive.py:117]: Oh yes, I see now, thanks!

File: zuko/flows/autoregressive.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (adrianjav) [reply zuko/flows/autoregressive.py:172]: Yes, I did misunderstand it, my bad. 
If that's the case, we could also compute the order in a pythonic way by just checking each variable in which generation level is:
```python
self.order = torch.tensor(
    [i for j in range(adjacency.size(0)) for i, x in enumerate(all_generations) if j in x]
)
```
But of course it is your call, I have no strong opinions on this matter.

File: zuko/flows/autoregressive.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrianjav) [reply zuko/flows/autoregressive.py:131]: That is actually a pretty good reason to checking it. I'll raise an error then 👍🏼

File: zuko/flows/autoregressive.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (francois-rozet) [reply zuko/flows/autoregressive.py:172]: I would prefer setting `self.order` to `None`. It is useless anyway.

**Comment by adrianjav** [comment]
Thanks a lot for the review. I think all changes are more or less clear and I will start implementing them. I actually would appreciate if, after the next changes, you could put the final touches (e.g., let the `__repr__` function to your liking) as it may be quicker that way.

PS: I am going on vacation tomorrow for a week, so I will be a bit less active.

**Comment by francois-rozet** [comment]
Sure, I'll finish the PR! Tell me when I can take over.

**Comment by adrianjav** [comment]
I just made most of the changes suggested in the review. As I said, I'd be grateful if you can take over and implement the string representation at your liking.

Another comment, with regards to the diagonal of `adjacency`. It should be zero for the work to code, but I think the reason you gave is a super valid one, so I have opted to forced the user to fill it with ones, and then I fill with zeroes. Feel free to change it if you think it is better doing it another way.

**Comment by adrianjav** [comment]
I think you can take over already :)

**Comment by francois-rozet** [comment]
Hello @adrianjav, I have made a few modifications to `__init__`, `extra_repr` and the tests. For the tests, I replaced the fixed adjacency matrices with random ones, which can help to detect more issues. For the representation, I chose to display the number of passes when the adjacency is provided.

For the `__init__`, I chose to revert to zeros on the diagonal as this is the convention in your paper and in [Wehenkel et. al (2021)](https://arxiv.org/abs/2006.02548).

I also really liked the way you added columns to the adjacency for the context and duplicated rows for the output so I made this part common for both `order` and `adjacency`.

**Comment by francois-rozet** [comment]
In the end I have reverted to ones on the diagonal. As you said, I think it is easier to understand that `adjacency` describes the transformation graph rather than the conditioning graph.

**Comment by adrianjav** [comment]
Glad to contribute! ☺️
